### DISCUSSION TITLE
Parent instance <Company at 0x7f0475536f10> is not bound to a Session; lazy load operation of attribute 'addresses'

### DISCUSSION THREAD

**Discussion by SanskarSans** [sqlalchemy/sqlalchemy#7780] [Usage Questions]
I have a table `Company` and `addresses`. When creating a company, a company object is created and then company user and lastly address is created with newly created company id. There is no issue in insertion, however when I try to access `company.addresses` I get above issue. I saw joinedload fixes the error but in my case I need that once insertion is successful and return the created company object with addresses. Does that mean after the company object, address object are committed to database should I again query something like this way `select(Company).join(Company.addresses).options(joinedload(Company.addresses)).filter(Company.id == company.id)` 

```python
class Company(Base):
    id: uuid.UUID = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4())
    owner_id: uuid.UUID = Column(UUID(as_uuid=True), ForeignKey('user.id', ondelete='CASCADE'))
    name: str = Column(String(200), nullable=False, index=True)
    is_active: bool = Column(Boolean, default=False)
    addresses: List[Address] = relationship('Address', back_populates='company')
    posts: List[Post] = relationship('Post', back_populates='company')

class Address(Base):
    id: uuid.UUID = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4())
    address_1: str = Column(String(100))
    address_2: str = Column(String(100))
    country: str = Column(String(100))
    city: str = Column(String(100))
    zipcode: str = Column(String(20))
    company_id: uuid.UUID = Column(UUID(as_uuid=True), ForeignKey('company.id'))
    company = relationship('Company', back_populates='addresses')


In my case, I need to insert in 3 tables while creating a company. 

async def company_create(info: Info, input: CompanyCreateInput) -> CompanyCreateResult:
    try:
        address = input.address
        input_without_address = dict(filter(lambda x: x[0] != 'address', input.__dict__.items())) # needs improvement
        data = CompanyCreate(owner_id=user.id, **input_without_address, address=address.__dict__)
    except ValidationError as err:
        error_msg = [Error(loc=str(e['loc']), msg = str(e['msg'])) for e in json.loads(err.json())]
        return CompanyCreateError(error=error_msg)
    async with async_session() as session:
        if await company_by_name(session=session, name=data.name) is not None:
            return CompanyCreateError(error=[Error(loc="name", msg="Name already exists.")])
            company, address = await create_company(input, owner_id=user.id, address=input.address, session=session)
            print("company", company.addresses) # throws error
        print('company, address', company.id)
    return CompanyCreateSuccess(company=cast(CompanyType, company))


async def create_company(company_input: CompanyCreate, owner_id, address, session: AsyncSession) -> Optional[Company]:
    input_without_address = dict(filter(lambda x: x[0] != 'address', company_input.__dict__.items()))
    company = Company(**input_without_address, owner_id=owner_id)
    session.add(instance=company)
    await session.flush()
    company_user = CompanyUser(user_id=owner_id, company_id=company.id)
    session.add(instance=company_user)
    await session.flush()
    address = Address(**address.__dict__, company_id=company.id)
    session.add(instance=address)
    await session.commit()
    return company
```
  How would you approach such issue?

**Comment by zzzeek** [comment]
well you can't use lazy loading with asyncio in any case, so if you want to pass around a Company object and access the addresses collection you would need to ensure it's been loaded in some way.  there's advice on this here: https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#preventing-implicit-io-when-using-asyncsession 

the code you show though shouldn't be throwing the error you describe:

```py
  async with async_session() as session:
        if await company_by_name(session=session, name=data.name) is not None:
            return CompanyCreateError(error=[Error(loc="name", msg="Name already exists.")])
            company, address = await create_company(input, owner_id=user.id, address=input.address, session=session)
            print("company", company.addresses) # throws error
```

that "company" object is bound to the session you are giving it.  it will raise an asyncio/ greenlet related error in the above case if I'm reading correctly but I dont see where it would see "company" as not bound to a session.

**Reply by SanskarSans** [reply]
Thank you very much for responding to my question with the direction. Like you said, I am getting asyncio error now. "Error: greenlet_spawn has not been called; can't call await_() here. Was IO attempted in an unexpected place?". I got confused using .statement like in the docs still gave me same error. 
```
company = await create_company(input, owner_id=user.id, address=input.address, session=session)
addresses = (await session.scalars(company.addresses.statement)).all()
```

**Reply by SanskarSans** [reply]
I did not understand how can I return addresses object as soon as company is created. However, I tried with just querying and it worked. I used `contains_eager` in this way

```python
async def companies(info:Info) -> List[Optional[CompanyType]]:
    user = await info.context.user()
    async with async_session() as session: 
        result = select(Company).join(Company.addresses).options(contains_eager(Company.addresses)).filter(Company.owner_id == user.id)
        # result = select(Company).filter(Company.owner_id == user.id)
        qs = (await session.execute(result)).scalars().unique()
    return qs
```

@zzzeek can you enlighten me more on how can I return address object along with company data like in `companies` function I have done but as soon as new company is created? Thank you in advance.

**Comment by Kevin-Candia** [comment]
I have the same error and I don't understand what is causing the problem

**Reply by zzzeek** [reply]
if you have the error message that's actually in the title of this discussion, the message means your object is [detached](https://docs.sqlalchemy.org/en/14/glossary.html#term-detached) from any Session.  the exception message itself should contain a hyperlink that points you to [this FAQ / error entry](https://docs.sqlalchemy.org/en/14/errors.html#parent-instance-x-is-not-bound-to-a-session-lazy-load-deferred-load-refresh-etc-operation-cannot-proceed) which should cover the full range of what this message means.
